File: splitsy-frontend/src/domain/shared/value-objects/Name.ts
import { ValueObject } from "./ValueObject";

export interface NameProps {
  value: string;
}

export class Name extends ValueObject<NameProps> {
  private constructor(props: NameProps) {
    super(props);
  }

  public static create(name: string): Name {
    if (!name || name.trim().length === 0) {
      throw new Error('Name is required');
    }
    if (name.trim().length < 2) {
      throw new Error('Name must be at least 2 characters');
    }
    if (name.trim().length > 100) {
      throw new Error('Name must be at most 100 characters');
    }
    return new Name({ value: this.format(name) })
  }

  public get value(): string {
    return this.props.value;
  }

  public static format(name: string): string {
    return name.trim();
  }
}

File: splitsy-backend/src/domain/expenses/value_objects/name.py
from dataclasses import dataclass

from src.domain.common.errors import ValidationError
from src.domain.common.value_objects import ValueObject


@dataclass(frozen=True)
class NameProps:
    value: str


class Name(ValueObject[NameProps]):
    def __init__(self, props: NameProps) -> None:
        super().__init__(props)
        self._value = props.value

    @property
    def value(self) -> str:
        return self._value

    @staticmethod
    def create(value: str) -> "Name":
        if not isinstance(value, str):
            raise ValidationError(f"Invalid name format: {value}")

        if len(value) < 1:
            raise ValidationError("Name should have at least 1 character")

        if len(value) > 100:
            raise ValidationError("Name should have at most 100 characters")

        return Name(NameProps(value=value))
